docs/CppInteroperability/Roadmap.md

```diff
@@ -0,0 +1,141 @@
+
+**Owned types** “own” some storage which can be copied and destroyed. An owned type must be copyable and destructible. The copy constructor must copy any storage that is owned by the type and the destructor must destroy that storage. Copies and destroys must balance out and these operations must not have side effects. Examples of owned types include `std::vector` and `std::string`. 
+
+**An iterator** represents a point in a range. Iterator types must provide a comparison operator and increment operator. Iterators can be returned by `begin` and `end` methods to form a range.
+
+**Trivial types** can be copied by copying the bits of a value of the trivial type and do not need any special destruction logic. Trivial types own their storage.
+
+**Projections** are references to storage. Projections of owned types are usually unsafe. Most iterators are projections. 
+
+
+### Reference types
+
+Swift types fall into two categories: value types and reference types. The same is true for Objective-C types, which makes importing types easy: structs are imported as value types and Objective-C classes are imported as Swift class types. The same is not true for C++. In typical C++ fashion, there is no clear idiom for defining reference types. All types in C++ are declared in the same way, but some types have reference semantics and others don’t. To be able to express reference types in Swift natively, users must annotate their reference types as such, which will tell the compiler to import them as Swift classes (which have the same semantics). 
+
+When importing a reference type, there are three main sub-categories that reference type can fall into: 1) **immortal** reference types are “intentionally leaked” and live for the duration of the program 2) **manually managed** reference types are constructed and destroyed manually by the programmer, these are unsafe reference types and should maybe be renamed accordingly 3) **automatically managed** reference types provide retain and release operations which are used to safely destroy a value of the type after its last use. 
+
+Reference types, their instances, methods on reference types, and other APIs that use reference types generally fit well into the existing Swift model and their use should be allowed without restriction. While it is possible to define unsafe APIs that use reference types, these APIs will not be any less safe than their Swift or C++ counterpart, so there is no reason to dis-allow them. Note: the only case where C++ APIs using reference types must be dis-allowed is when there is not one level of indirection provided (either a reference or pointer). In this case, the C++ API is not using the type as a reference types, and thus breaks the reference type definition (above).
```

Comment:
```suggestion
Reference types, their instances, methods on reference types, and other APIs that use reference types generally fit well into the existing Swift model and their use should be allowed without restriction. While it is possible to define unsafe APIs that use reference types, these APIs will not be any less safe than their Swift or C++ counterpart, so there is no reason to disallow them. Note: the only case where C++ APIs using reference types must be disallowed is when there is not one level of indirection provided (either a reference or pointer). In this case, the C++ API is not using the type as a reference types, and thus breaks the reference type definition (above).
```

Reply:
I'm not sure I understand the case you are thinking of with "there is not one level of indirection provided".  Perhaps an example would be helpful.

Reply:
Here's an example (I will include something like this with all other examples I'm putting together at the end). Maybe you can help me find a good way to describe this pattern, though?

```
struct SILFunction { ... } // reference type

SILFunction *SILBuilder::makeFunction(...) // Good API: uses SILFunction though one level of indirection
StringRef getName(SILFunction &fn) // Good API: uses SILFunction though one level of indirection

SILFunction findFunc() // Bad API: uses SILFunction as a value types and tries to copy it out
```

The final API here uses a reference type as a value type, making a copy. Does that make the "not one level of indirection" issue clear?

Reply:
Sorry, I'm still not following.  I'm reading the last "bad" example as:

```
void findFunc(SILFunction *);
```

which is an out parameter.  Is it about using reference types by reference?

Reply:
@compnerd can you confirm that the example I added towards the end of the document clears this up?